Fix problem with String.equals not checking the argument is a String TG-1619

romainbrenguier opened the pull request:
The previous implementation was assuming the given Object was a String, which could lead CBMC into mistakenly thinking objects of other types (Integer for instance) could be equal to a string.

romainbrenguier committed:
Fix String.equals to check for class identifier

romainbrenguier committed:
Test for String.equals with class identifier check

allredj reviewed (approved):
Looks fine, as far as I can tell.

romainbrenguier committed:
Prevent use of CharSequence as a class_identifier

CharSequence is abstract and should not appear as a class identifer.
Instead we use Sring which implements CharSequence.

smowton commented on src/java_bytecode/java_object_factory.cpp:
```diff
@@ -571,9 +571,15 @@ codet initialize_nondet_string_struct(
+  // @clsid = java::java.lang.String or similar
+  // Note that CharSequence cannot be used as classid because it's abstract
+  // \todo allow StringBuffer and StringBuilder as classid for Charsequence
+  const irep_idt &class_id =
```
Explain in the comment why you're allowing classes other than `String` to be aliased with `String` here?

smowton commented on src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -923,8 +922,30 @@ codet java_string_library_preprocesst::make_equals_function_code(
+    make_function_application(
+      ID_cprover_string_equal_func, args, type.return_type(), symbol_table)));
+
+  code.add(code_returnt(if_exprt(
```
`and_exprt`?

romainbrenguier replied:
The problem is that they don't have the same type: `arg_is_string` has `bool_type()` and `string_equals` has `java_boolean_type()`, these have different bitwidth, What I want in the end is something of type `java_boolean_type()`

smowton replied:
fair enough
